tools/tests/form-control/validation.test.js

```diff
@@ -0,0 +1,169 @@
+        currentElement.parentElement.removeChild(currentElement);
+    }
+
+    document.body.appendChild(el);
+
+    return new Promise(resolve => {
+        waitForStyles(resolve);
+    });
+}
+
+
+describe('Form validation', () => {
+    afterAll(() => {
+        const tooltips = document.querySelectorAll('gameface-tooltip');
+        for(let i = 0; i < tooltips.length; i++) {
+            tooltips[i].parentElement.removeChild(tooltips[i]);
+        }
+
+        let currentElement = document.querySelector('gameface-currentElement');
+
+        if (currentElement) {
+            currentElement.parentElement.removeChild(currentElement);
+        }
+    });
+
+    beforeEach(function (done) {
+        setupFormValidationTestPage().then(done);
+    });
+
+    it('Should show value missing error', async() => {
+        const input = document.querySelector('#role');
+        input.value = '';
+
+        click(document.querySelector('#submit'));
+
+        return createAsyncSpec(() => {
+            const tooltip = document.querySelector('#input-role-error-tooltip');
+            assert(tooltip.style.display !== 'none', 'Tooltip was not displayed!');
+
+            const expected = errorMessages.get('valueMissing');
+            const received = getTextContent(tooltip);;
+            tooltip.hide();
+            assert(received === expected, `The error message is not correct. Expected: ${expected}. Received: ${received}.`);
+        });
+    });
+
+    it('Should show value too long error', async() => {
+        const username = document.querySelector('#username');
+        username.value = 'This is too long';
+
+        click(document.querySelector('#submit'));
+
+        return createAsyncSpec(() => {
+            const tooltip = document.querySelector('#input-username-error-tooltip');
+            assert(tooltip.style.display !== 'none', 'Tooltip was not displayed!');
+
+            const expected = errorMessages.get('tooLong');
+            const received = getTextContent(tooltip);;
+            tooltip.hide();
+            assert(received === expected, `The error message is not correct. Expected: ${expected}. Received: ${received}.`);
+        });
+    });
+
+    it('Should show value too short error', async() => {
+        const username = document.querySelector('#username');
+        username.value = 'Th';
+
+        click(document.querySelector('#submit'));
+
+        return createAsyncSpec(() => {
+            const tooltip = document.querySelector('#input-username-error-tooltip');
+            assert(tooltip.style.display !== 'none', 'Tooltip was not displayed!');
+
+            const expected = errorMessages.get('tooShort');
+            const received = getTextContent(tooltip);;
+            tooltip.hide();
+            assert(received === expected, `The error message is not correct. Expected: ${expected}. Received: ${received}.`);
+        });
+    });
+
+    it('Should show value too small - range underflow.', async() => {

```

Comment:
This test isn't supposed to be "range underflow" case.
Create another test with input type number for that. Same for range overflow.
The password type is like the text type i.e. number of characters.

Reply:
It is supposed to be range underflow.
The test is supposed to test the validation methods.
I didn't use input type number as it is not supported in GF.
For now it is OK to use text input just to test the valueOverflow and valueUnderflow methods.
When we do the task about validating and serializing the text field we can use the correct input methods.
It is not a type password, but I see how the name might confuse you, I'll change it to age.